numpy/core/src/umath/loops.c.src

```diff
@@ -2720,46 +2722,239 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
+#if !@out_bool@
         PyObject **out = (PyObject **)op1;
-        PyObject *ret = NULL;
-        int v;
+#endif
+
+        in1 = in1 ? in1 : Py_None;
+
+        if (PyArray_CheckScalar(in1) || PyArray_Check(in1)) {
+            ret = PyObject_CallFunctionObjArgs(ufunc, in1, NULL);
+        }
+        else if (_is_float_complex(in1)) {
+            ret = PyObject_CallFunctionObjArgs(ufunc, in1, NULL);
+            if (ret != NULL) {
+                arr = (PyArrayObject*)PyArray_FromScalar(ret, NULL);
+                Py_DECREF(ret);
+                ret = PyArray_GETITEM(arr, PyArray_DATA(arr));
+                Py_XDECREF(arr);
+            }
+        }
+        else if ((attr = PyObject_GetAttrString(in1, "@func@")) != NULL && 
+                PyCallable_Check(attr)) {
+            ret = PyObject_CallFunctionObjArgs(attr, NULL);
+            Py_DECREF(attr);
+        }
+        else {
+            /* clear any error from attr lookup */
+            PyErr_Clear();
+            Py_XDECREF(attr);
+
+            ret = PyObject_CallFunctionObjArgs(fallback, in1, NULL);
+        }
 
-        if (in1 == NULL) {
-            in1 = Py_None;
+        if (ret == NULL) {
+            return;
         }
 
-        if ((v = PyObject_RichCompareBool(in1, zero, Py_LT)) == 1) {
-            ret = PyLong_FromLong(-1);
+#if @out_bool@
+        *((npy_bool *)op1) = PyObject_IsTrue(ret);
+        Py_DECREF(ret);
+#else
+        Py_XDECREF(*out);
+        *out = ret;
+#endif
+    }
+}
+/**end repeat**/
+
+/**begin repeat
+ * #func = logical_and, logical_or, logical_xor, maximum, minimum, fmax, fmin,
+           logaddexp, logaddexp2, copysign, ldexp, hypot, remainder, fmod,
+           nextafter#
+ */
+NPY_NO_EXPORT void
+OBJECT_@func@(char **args, npy_intp *dimensions, npy_intp *steps, void *data)
+{
+    PyObject *ret;
+    PyArrayObject *arr;
+    static PyObject *ufunc=NULL, *fallback=NULL, *attr=NULL;
+
+    /* variables for BINARY_LOOP */
+    char *ip1 = args[0], *ip2 = args[1], *op1 = args[2];
+    npy_intp is1 = steps[0], is2 = steps[1], os1 = steps[2];
+    npy_intp n = dimensions[0];
+    npy_intp i;
+
+    npy_cache_import("numpy", "@func@", &ufunc);
+    if (ufunc == NULL) {
+        return;
+    }
+    npy_cache_import("numpy.core._objectmath", "@func@", &fallback);
+    if (fallback == NULL) {
+        return;
+    }
+
+    for (i = 0; i < n; i++, ip1 += is1, ip2 += is2, op1 += os1) {
+        PyObject *in1 = *(PyObject **)ip1;
+        PyObject *in2 = *(PyObject **)ip2;
+        PyObject **out = (PyObject **)op1;
+
+        in1 = in1 ? in1 : Py_None;
+        in2 = in2 ? in2 : Py_None;
+        ret = NULL;
+    
+        if (_is_float_complex(in1) && _is_float_complex(in2)) {
+            ret = PyObject_CallFunctionObjArgs(ufunc, in1, in2, NULL);
+            if (ret != NULL) {
+                arr = (PyArrayObject*)PyArray_FromScalar(ret, NULL);
+                Py_DECREF(ret);
+                ret = PyArray_GETITEM(arr, PyArray_DATA(arr));
```

Comment:
For pure-python float and complex objects we call the ufunc on the object, which converts to a numpy type, eg float -> float64. Then we convert back to python types by calling `.item()`, but this method only exists on arrays so we need to convert the scalar to an array first, then call `PyArray_GETITEM`.